File: GSI QA testing tool/Finder.cs
﻿using OpenQA.Selenium;
using OpenQA.Selenium.Interactions;
using OpenQA.Selenium.Support.UI;
using SeleniumExtras.WaitHelpers;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Timers;
using System.Windows.Controls.Primitives;
using static System.Net.Mime.MediaTypeNames;

namespace GSI_QA_testing_tool
{
    public class Finder
    {

        public static bool FindIt(IWebDriver driver, String path)
        {
            return driver.FindElements(By.XPath(path)).Count != 0;        
        }


        public static void ClickIt(IWebDriver driver, String path)
        {

            if (driver.FindElements(By.XPath(path)).Count != 0)

            {
               driver.FindElement(By.XPath(path)).Click();
            }
        }


        public static void ClickIt(IWebDriver driver, String path, WebDriverWait wait)
        {
            if (driver.FindElements(By.XPath(path)).Count != 0)
            { 
            wait.Until(ExpectedConditions.ElementToBeClickable(By.XPath(path))).Click();
            wait.Until(ExpectedConditions.StalenessOf(driver.FindElement(By.XPath(path))));
            }
        }


        public static void WaitStaleClickIt(IWebDriver driver, String path, WebDriverWait wait)
        {
            if (driver.FindElements(By.XPath(path)).Count != 0)
            {
                wait.Until(ExpectedConditions.StalenessOf(driver.FindElement(By.XPath(path))));
                wait.Until(ExpectedConditions.ElementToBeClickable(By.XPath(path))).Click();
              
            }
        }


        public static void WaitClickableClickIt(IWebDriver driver, String path, WebDriverWait wait)
        {
            if (driver.FindElements(By.XPath(path)).Count != 0)
            {
                wait.Until(ExpectedConditions.ElementToBeClickable(By.XPath(path))).Click();
              
            }
        }



        public static void SendText(IWebDriver driver, string path, string text)
        {
            if (driver.FindElements(By.XPath(path)).Count != 0)

            {
                driver.FindElement(By.XPath(path)).SendKeys(text);
            }
        }

        public static void SendText(IWebDriver driver, string path, string text, WebDriverWait wait)
        {
            if (driver.FindElements(By.XPath(path)).Count != 0)

            {
                driver.FindElement(By.XPath(path)).SendKeys(text);
                wait.Until(ExpectedConditions.StalenessOf(driver.FindElement(By.XPath(path))));
            }
        }



        public static void WaitStaleSendText(IWebDriver driver, string path, string text, WebDriverWait wait)
        {
            if (driver.FindElements(By.XPath(path)).Count != 0)

            {
                wait.Until(ExpectedConditions.StalenessOf(driver.FindElement(By.XPath(path))));
                driver.FindElement(By.XPath(path)).SendKeys(text);
            }
        }




        public static void ActionsSendText(IWebDriver driver, String path, String text)
        {
            if (driver.FindElements(By.XPath(path)).Count != 0)
            {
                Actions act = new Actions(driver);
                act.MoveToElement(driver.FindElement(By.XPath(path)));
                act.Click();
                act.SendKeys(text);
                act.Pause(new TimeSpan(7000));
                act.SendKeys(Keys.Down);
                act.Pause(new TimeSpan(1000));
                act.SendKeys(Keys.Enter);
                act.Pause(new TimeSpan(1000));
                act.Build().Perform();
            }
        }



        public static void UseAlert(IWebDriver driver, int time)
        {
            try
            {
                IAlert alert = driver.SwitchTo().Alert();
                Thread.Sleep(time);
                alert.Accept();
            }
            catch (NoAlertPresentException nap)
            {
                Debug.WriteLine(nap.Message);
            }
        }


        public static void WaitDocToLoad(IWebDriver driver, WebDriverWait wait)
        {
            wait.Until(driver1 => ((IJavaScriptExecutor)driver).ExecuteScript("return document.readyState").Equals("complete"));
        }





        public static void UseDropDownByValue(IWebDriver driver, String path, String value)
        {
            if (driver.FindElements(By.XPath(path)).Count != 0)
            {
                SelectElement select = new SelectElement(driver.FindElement(By.XPath(path)));
                select.SelectByValue(value);
            }
        }



        public static void UseDropDownByValue(IWebDriver driver, string path, string value, WebDriverWait wait)
        {
            if (driver.FindElements(By.XPath(path)).Count != 0)
            {
                SelectElement select = new SelectElement(driver.FindElement(By.XPath(path)));
                select.SelectByValue(value);
                wait.Until(ExpectedConditions.StalenessOf(driver.FindElement(By.XPath(path))));
            }
        }


        public static void UseDropDownByText(IWebDriver driver, string path, string value)
        {
            if (driver.FindElements(By.XPath(path)).Count != 0)
            {
                SelectElement select = new SelectElement(driver.FindElement(By.XPath(path)));
                select.SelectByText(value);
            }
        }

        public static void UseDropDownByText(IWebDriver driver, string path, string value, WebDriverWait wait)
        {
            if (driver.FindElements(By.XPath(path)).Count != 0)
            {
                SelectElement select = new SelectElement(driver.FindElement(By.XPath(path)));
                select.SelectByText(value);
                wait.Until(ExpectedConditions.StalenessOf(driver.FindElement(By.XPath(path))));
            }
        }


        public static void UseDropDownByIndex(IWebDriver driver, string path, int value)
        {
            if (driver.FindElements(By.XPath(path)).Count != 0)
            {
                SelectElement select = new SelectElement(driver.FindElement(By.XPath(path)));
                select.SelectByIndex(value);
            }
        }


        public static void UseDropDownByIndex(IWebDriver driver, string path, int value, WebDriverWait wait)
        {
            if (driver.FindElements(By.XPath(path)).Count != 0)
            {
                SelectElement select = new SelectElement(driver.FindElement(By.XPath(path)));
                select.SelectByIndex(value);
                wait.Until(ExpectedConditions.StalenessOf(driver.FindElement(By.XPath(path))));
            }
        }



 








    }
}

File: GSI QA testing tool/Finder.vb
Imports OpenQA.Selenium
Imports OpenQA.Selenium.Interactions
Imports OpenQA.Selenium.Support.UI
Imports SeleniumExtras.WaitHelpers
Imports System
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.Linq
Imports System.Text
Imports System.Threading
Imports System.Threading.Tasks
Imports System.Timers
Imports System.Windows.Controls.Primitives
Imports System.Net.Mime.MediaTypeNames

Namespace GSI_QA_testing_tool
    Public Class Finder
        Public Shared Function FindIt(ByVal driver As IWebDriver, ByVal path As String) As Boolean
            Return driver.FindElements(By.XPath(path)).Count <> 0
        End Function

        Public Shared Function FindItByCSS(ByVal driver As IWebDriver, ByVal CSSPath As String) As Boolean
            Return driver.FindElements(By.CssSelector(CSSPath)).Count <> 0
        End Function


        Public Shared Sub JSClickIt(ByVal driver As IWebDriver, ByVal CSSpath As String, ByVal js As IJavaScriptExecutor)
            If driver.FindElements(By.CssSelector(CSSpath)).Count <> 0 Then
                js.ExecuteScript("arguments[0].click();", driver.FindElement(By.CssSelector(CSSpath)))
            End If
        End Sub

        Public Shared Sub ClickIt(ByVal driver As IWebDriver, ByVal path As String)
            If driver.FindElements(By.XPath(path)).Count <> 0 Then
                driver.FindElement(By.XPath(path)).Click()
            End If
        End Sub

        Public Shared Sub ClickIt(ByVal driver As IWebDriver, ByVal path As String, ByVal wait As WebDriverWait)
            If driver.FindElements(By.XPath(path)).Count <> 0 Then
                wait.Until(ExpectedConditions.ElementToBeClickable(By.XPath(path))).Click()
                wait.Until(ExpectedConditions.StalenessOf(driver.FindElement(By.XPath(path))))
            End If
        End Sub

        Public Shared Sub WaitStaleClickIt(ByVal driver As IWebDriver, ByVal path As String, ByVal wait As WebDriverWait)
            If driver.FindElements(By.XPath(path)).Count <> 0 Then
                wait.Until(ExpectedConditions.StalenessOf(driver.FindElement(By.XPath(path))))
                wait.Until(ExpectedConditions.ElementToBeClickable(By.XPath(path))).Click()
            End If
        End Sub

        Public Shared Sub WaitClickableClickIt(ByVal driver As IWebDriver, ByVal path As String, ByVal wait As WebDriverWait)
            If driver.FindElements(By.XPath(path)).Count <> 0 Then
                wait.Until(ExpectedConditions.ElementToBeClickable(By.XPath(path))).Click()
            End If
        End Sub

        Public Shared Sub SendText(ByVal driver As IWebDriver, ByVal path As String, ByVal text As String)
            If driver.FindElements(By.XPath(path)).Count <> 0 Then
                driver.FindElement(By.XPath(path)).SendKeys(text)
            End If
        End Sub

        Public Shared Sub SendText(ByVal driver As IWebDriver, ByVal path As String, ByVal text As String, ByVal wait As WebDriverWait)
            If driver.FindElements(By.XPath(path)).Count <> 0 Then
                driver.FindElement(By.XPath(path)).SendKeys(text)
                wait.Until(ExpectedConditions.StalenessOf(driver.FindElement(By.XPath(path))))
            End If
        End Sub

        Public Shared Sub WaitStaleSendText(ByVal driver As IWebDriver, ByVal path As String, ByVal text As String, ByVal wait As WebDriverWait)
            If driver.FindElements(By.XPath(path)).Count <> 0 Then
                wait.Until(ExpectedConditions.StalenessOf(driver.FindElement(By.XPath(path))))
                driver.FindElement(By.XPath(path)).SendKeys(text)
            End If
        End Sub

        Public Shared Sub WaitClickableSendText(ByVal driver As IWebDriver, ByVal path As String, ByVal text As String, ByVal wait As WebDriverWait)
            If driver.FindElements(By.XPath(path)).Count <> 0 Then
                wait.Until(ExpectedConditions.ElementToBeClickable(By.XPath(path)))
                driver.FindElement(By.XPath(path)).SendKeys(text)
            End If
        End Sub

        Public Shared Sub ActionsSendText(ByVal driver As IWebDriver, ByVal path As String, ByVal text As String, ByVal Optional wait As WebDriverWait = Nothing)
            If driver.FindElements(By.XPath(path)).Count <> 0 Then
                Dim act As Actions = New Actions(driver)
                act.MoveToElement(driver.FindElement(By.XPath(path)))
                act.Click()
                act.Pause(New TimeSpan(300))
                act.SendKeys(text)
                act.DoubleClick()
                act.Pause(New TimeSpan(1000))
                act.SendKeys(Keys.Down)
                act.SendKeys(Keys.Down)
                act.SendKeys(Keys.Enter)
                act.Build().Perform()
            End If

            If wait IsNot Nothing Then
                wait.Until(ExpectedConditions.StalenessOf(driver.FindElement(By.XPath(path))))
            End If
        End Sub

        Public Shared Sub UseAlert(ByVal driver As IWebDriver, ByVal time As Integer)
            Try
                Dim alert As IAlert = driver.SwitchTo().Alert()
                Thread.Sleep(time)
                alert.Accept()
            Catch nap As NoAlertPresentException
                Debug.WriteLine(nap.Message)
            End Try
        End Sub

        Public Shared Sub UseDropDownByValue(ByVal driver As IWebDriver, ByVal path As String, ByVal value As String)
            If driver.FindElements(By.XPath(path)).Count <> 0 Then
                Dim [select] As SelectElement = New SelectElement(driver.FindElement(By.XPath(path)))
                [select].SelectByValue(value)
            End If
        End Sub

        Public Shared Sub UseDropDownByValue(ByVal driver As IWebDriver, ByVal path As String, ByVal value As String, ByVal wait As WebDriverWait)
            If driver.FindElements(By.XPath(path)).Count <> 0 Then
                Dim [select] As SelectElement = New SelectElement(driver.FindElement(By.XPath(path)))
                [select].SelectByValue(value)
                wait.Until(ExpectedConditions.StalenessOf(driver.FindElement(By.XPath(path))))
            End If
        End Sub

        Public Shared Sub UseDropDownByText(ByVal driver As IWebDriver, ByVal path As String, ByVal value As String)
            If driver.FindElements(By.XPath(path)).Count <> 0 Then
                Dim [select] As SelectElement = New SelectElement(driver.FindElement(By.XPath(path)))
                [select].SelectByText(value)
            End If
        End Sub

        Public Shared Sub UseDropDownByText(ByVal driver As IWebDriver, ByVal path As String, ByVal value As String, ByVal wait As WebDriverWait)
            If driver.FindElements(By.XPath(path)).Count <> 0 Then
                Dim [select] As SelectElement = New SelectElement(driver.FindElement(By.XPath(path)))
                [select].SelectByText(value)
                wait.Until(ExpectedConditions.StalenessOf(driver.FindElement(By.XPath(path))))
            End If
        End Sub

        Public Shared Sub UseDropDownByIndex(ByVal driver As IWebDriver, ByVal path As String, ByVal value As Integer)
            If driver.FindElements(By.XPath(path)).Count <> 0 Then
                Dim [select] As SelectElement = New SelectElement(driver.FindElement(By.XPath(path)))
                [select].SelectByIndex(value)
            End If
        End Sub

        Public Shared Sub UseDropDownByIndex(ByVal driver As IWebDriver, ByVal path As String, ByVal value As Integer, ByVal wait As WebDriverWait)
            If driver.FindElements(By.XPath(path)).Count <> 0 Then
                Dim [select] As SelectElement = New SelectElement(driver.FindElement(By.XPath(path)))
                [select].SelectByIndex(value)
                wait.Until(ExpectedConditions.StalenessOf(driver.FindElement(By.XPath(path))))
            End If
        End Sub
    End Class
End Namespace
